src/items/enumerations.md

```diff
@@ -122,6 +167,59 @@ enum OverflowingDiscriminantError2 {
 }
 ```
 
+### Accessing Discriminant
+
+#### Via `mem::discriminant`
+
+[`mem::discriminant`] returns an opaque reference to the discriminant of
+an enum value which can be compared. This cannot be used to get the value
+of the discriminant.
+
+#### Casting
+
+If an enumeration is [unit-only] (with no tuple and struct variants), then its
+discriminant can be directly accessed with a [numeric cast]; e.g.:
+
+```rust
+enum Enum {
+    Foo,
+    Bar,
+    Baz,
+}
+
+assert_eq!(0, Enum::Foo as isize);
+assert_eq!(1, Enum::Bar as isize);
+assert_eq!(2, Enum::Baz as isize);
+```
+
+#### Pointer Casting
+
+If the enumeration specifies a [primitive representation], then the
+discriminant may be reliably accessed via unsafe pointer casting:
+
+```rust
+#[repr(u8)]
+enum Enum {
+    Unit,
+    Tuple(bool),
+    Struct{a: bool},
+}
+
+impl Enum {
+    fn discriminant(&self) -> u8 {
+        unsafe { *(self as *const Self as *const u8) }
```

Comment:
This doesn't look right to me?  Just because it's `repr(u8)` I don't think that means that the discriminant is necessarily *first*.

If the `enum` were `repr(C)` or `repr(C, u8)` *then* it'd be ok because the field order would be specified.  But for `repr(Rust)` I'm pretty sure we're allowed to put the discriminant *after* the payload, if we wanted to.

Reply:
Oh, never mind, it looks like <https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html#guide-level-explanation> means that `repr(u8)` disables all field reordering too, not just sets the discriminant size :(

Reply:
I found that odd, too. The RFC justifies this by saying the discriminant position needs to be deterministic, which seems possible to do without disabling other field reordering?

Reply:
(At this point it seems too late to change, but an edition change could require `repr(C)` for deterministic field ordering.)

Reply:
That would need another flag to distinguish between the union-of-structs and struct-containing-union-of-structs layouts, though.  Which might be nice, though!

I suppose the argument is that the size of the discriminant can't matter to safe code, so the only point of ever specifying its width is to get deterministic layout?